### PR TITLE
bug fixes in python

### PR DESCRIPTION
**Pull request by paulfantom**
Do not merge yet.

1. fixed mysql issue `invalid literal for int() with base 10: 'absolute'`

### CODE REVIEW

**Comment by paulfantom** [comment]
I cannot replicate issue with `squid`.

**Comment by ktsaou** [comment]
gentoo

**Comment by paulfantom** [comment]
Does `squid.chart.py` print lines in netdata dashboard as well as logs error messages about lacking dimensions?

**Comment by ktsaou** [comment]
this works

**Comment by ktsaou** [comment]
I don't see any issues anywhere. Merge?

**Comment by paulfantom** [comment]
Test now. I had to reimplement socket timeout. Check if you see double update_every time (I don't see it on fedora and on debian)

**Comment by ktsaou** [comment]
```
$ ./plugins.d/python.d.plugin redis
 INFO: Using python v2
python.d INFO: reading configuration file: /etc/netdata/python.d.conf
python.d INFO: MODULES_DIR='/data/src/netdata-paulfantom.git/python.d/', CONFIG_DIR='/etc/netdata/', UPDATE_EVERY=1, ONLY_MODULES=['redis']
python.d ERROR: redis_socket No request specified. Using: 'INFO
'
python.d ERROR: redis/local: is misbehaving. Reason:sequence item 1: expected string, error found
python.d ERROR: redis_tcp No unix socket specified. Trying TCP/IP socket.
python.d ERROR: redis_tcp No request specified. Using: 'INFO
'
python.d ERROR: redis_tcp_local No data received
python.d ERROR: redis/local: check() function reports failure.
python.d FATAL: no more jobs
DISABLE
```

**Comment by ktsaou** [comment]
apache works
squid works
sensors works
mysql works

**Comment by ktsaou** [comment]
```
$ ./plugins.d/python.d.plugin redis
 INFO: Using python v2
python.d INFO: reading configuration file: /etc/netdata/python.d.conf
python.d INFO: MODULES_DIR='/data/src/netdata-paulfantom.git/python.d/', CONFIG_DIR='/etc/netdata/', UPDATE_EVERY=1, ONLY_MODULES=['redis']
python.d ERROR: redis_socket No request specified. Using: 'INFO
'
python.d ERROR: redis_socket_local [Errno 2] No such file or directory
python.d ERROR: redis_socket_local No data received
python.d ERROR: redis/local: check() function reports failure.
python.d ERROR: redis_tcp No unix socket specified. Trying TCP/IP socket.
python.d ERROR: redis_tcp No request specified. Using: 'INFO
'
python.d ERROR: redis_tcp_local No data received
python.d ERROR: redis/local: check() function reports failure.
python.d FATAL: no more jobs
DISABLE
```

**Comment by ktsaou** [comment]
2 modules set a default update_every at their code.
Could you please comment it out?

- example
- mysql

**Comment by ktsaou** [comment]
You spent your whole day on it...

**Comment by ktsaou** [comment]
Thanks anyway! Your work is very important!

**Comment by paulfantom** [comment]
Not the first day and probably not the last one.

I think it works. Check yourself (mainly `squid` and `redis` since they use the same class `SocketService`).

**Comment by ktsaou** [comment]
redis works.

squid does not:

```
 $ ./plugins.d/python.d.plugin squid
 INFO: Using python v2
python.d INFO: reading configuration file: /etc/netdata/python.d.conf
python.d INFO: MODULES_DIR='/data/src/netdata-paulfantom.git/python.d/', CONFIG_DIR='/etc/netdata/', UPDATE_EVERY=1, ONLY_MODULES=['squid']
python.d ERROR: squid_tcp3 No unix socket specified. Trying TCP/IP socket.
python.d ERROR: squid_tcp3 No data returned
python.d ERROR: squid: check() function reports failure.
python.d ERROR: squid_tcp2 No unix socket specified. Trying TCP/IP socket.
python.d ERROR: squid_tcp2 [Errno 111] Connection refused
python.d ERROR: squid_tcp2 No data returned
python.d ERROR: squid: check() function reports failure.
python.d ERROR: squid_tcp1 No unix socket specified. Trying TCP/IP socket.
python.d ERROR: squid_tcp1 No data returned
python.d ERROR: squid: check() function reports failure.
python.d ERROR: squid_tcp4 No unix socket specified. Trying TCP/IP socket.
python.d ERROR: squid_tcp4 [Errno 111] Connection refused
python.d ERROR: squid_tcp4 No data returned
python.d ERROR: squid: check() function reports failure.
python.d FATAL: no more jobs
DISABLE
```

**Comment by ktsaou** [comment]
Retrying it worked:

```
$ ./plugins.d/python.d.plugin squid
 INFO: Using python v2
python.d INFO: reading configuration file: /etc/netdata/python.d.conf
python.d INFO: MODULES_DIR='/data/src/netdata-paulfantom.git/python.d/', CONFIG_DIR='/etc/netdata/', UPDATE_EVERY=1, ONLY_MODULES=['squid']
python.d ERROR: squid_tcp3 No unix socket specified. Trying TCP/IP socket.
python.d ERROR: squid_tcp2 No unix socket specified. Trying TCP/IP socket.
python.d ERROR: squid_tcp2 [Errno 111] Connection refused
python.d ERROR: squid_tcp2 No data returned
python.d ERROR: squid: check() function reports failure.
python.d ERROR: squid_tcp1 No unix socket specified. Trying TCP/IP socket.
python.d ERROR: squid_local already exists. Created with ''
CHART squid_local.clients_net '' 'Squid Client Bandwidth' kilobits/s clients squid.clients.net area 60000 1
DIMENSION client_http_kbytes_in in incremental 8 1
DIMENSION client_http_kbytes_out out incremental -8 1
DIMENSION client_http_hit_kbytes_out hits incremental -8 1
CHART squid_local.clients_requests '' 'Squid Client Requests' requests/s clients squid.clients.requests line 60001 1
DIMENSION client_http_requests requests incremental 1 1
DIMENSION client_http_hits hits incremental 1 1
DIMENSION client_http_errors errors incremental -1 1
CHART squid_local.servers_net '' 'Squid Server Bandwidth' kilobits/s servers squid.servers.net area 60002 1
DIMENSION server_all_kbytes_in in incremental 8 1
DIMENSION server_all_kbytes_out out incremental -8 1
CHART squid_local.servers_requests '' 'Squid Server Requests' requests/s servers squid.servers.requests line 60003 1
DIMENSION server_all_requests requests incremental 1 1
DIMENSION server_all_errors errors incremental -1 1

CHART netdata.plugin_pythond_squid_local '' 'Execution time for squid_local plugin' 'milliseconds / run' python.d netdata.plugin_python area 145000 1
DIMENSION run_time 'run time' absolute 1 1

python.d ERROR: squid_tcp4 timed out
python.d ERROR: squid/tcp4: create() function reports failure.
BEGIN squid_local.clients_net 3011662
SET client_http_kbytes_in = 206905
SET client_http_kbytes_out = 13945896
SET client_http_hit_kbytes_out = 208178
END
BEGIN squid_local.clients_requests 3011662
SET client_http_requests = 282589
SET client_http_hits = 7586
SET client_http_errors = 5125
END
BEGIN squid_local.servers_net 3011662
SET server_all_kbytes_in = 13383304
SET server_all_kbytes_out = 182538
END
BEGIN squid_local.servers_requests 3011662
SET server_all_requests = 84662
SET server_all_errors = 0
END

BEGIN netdata.plugin_pythond_squid_local 3011662
SET run_time = 5
END
BEGIN squid_local.clients_net 194870
SET client_http_kbytes_in = 206905
SET client_http_kbytes_out = 13945897
SET client_http_hit_kbytes_out = 208178
END
BEGIN squid_local.clients_requests 194870
SET client_http_requests = 282590
SET client_http_hits = 7586
SET client_http_errors = 5125
END
BEGIN squid_local.servers_net 194870
SET server_all_kbytes_in = 13383304
SET server_all_kbytes_out = 182538
END
BEGIN squid_local.servers_requests 194870
SET server_all_requests = 84662
SET server_all_errors = 0
END

BEGIN netdata.plugin_pythond_squid_local 194870
SET run_time = 6
END
BEGIN squid_local.clients_net 3008138
SET client_http_kbytes_in = 206905
SET client_http_kbytes_out = 13945902
SET client_http_hit_kbytes_out = 208178
END
BEGIN squid_local.clients_requests 3008138
SET client_http_requests = 282593
SET client_http_hits = 7586
SET client_http_errors = 5125
END
BEGIN squid_local.servers_net 3008138
SET server_all_kbytes_in = 13383304
SET server_all_kbytes_out = 182538
END
BEGIN squid_local.servers_requests 3008138
SET server_all_requests = 84662
SET server_all_errors = 0
END

```

something strange is happening...

**Comment by ktsaou** [comment]
I cannot make it work again...

**Comment by paulfantom** [comment]
I got the same. It's either redis working or squid, never both. However this version is closest to making them both work.

I am using socket in non-blocking mode, but somehow when I send and immediately recive bytes it sometime says that resource is temporarily unavailable.

**Comment by ktsaou** [comment]
Why don't you pass an option from each module to enable / disable non-blocking mode?

**Comment by ktsaou** [comment]
So one module will have it in blocking and the other in non-blocking mode.

File: python.d/python_modules/base.py [hunk 1]
```diff
@@ -512,23 +512,31 @@ def _get_raw_data(self):
...
-                    buf = sock.recv(1024, 0x40)  # get 1024 bytes in NON-BLOCKING mode
-                except socket.error:
-                    break
-
-                if len(buf) == 0:
-                    break
-                else:
-                    data += buf
-        except:
+            data = sock.recv(65535, 0x40).decode()
```
Comment (ktsaou) [inline python.d/python_modules/base.py]: what is 0x40?

File: python.d/python_modules/base.py [hunk 2]
```diff
@@ -488,15 +486,17 @@ def _get_raw_data(self):
                 if self.unix_socket is None:
                     sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
                     sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
-                    #sock.settimeout(self.update_every)
+                    #sock.setsockopt(socket.SOL_SOCKET, socket.TCP_NODELAY, 1)
+                    #sock.setsockopt(socket.SOL_SOCKET, socket.SO_KEEPALIVE, 1)
```
Comment (ktsaou) [inline python.d/python_modules/base.py]: This keep-alive is different from `Connection: keep-alive` in HTTP.

File: python.d/python_modules/base.py [hunk 3]
```diff
@@ -488,15 +486,17 @@ def _get_raw_data(self):
                 if self.unix_socket is None:
                     sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
                     sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
-                    #sock.settimeout(self.update_every)
+                    #sock.setsockopt(socket.SOL_SOCKET, socket.TCP_NODELAY, 1)
```
Comment (ktsaou) [inline python.d/python_modules/base.py]: This may save a few ms per request

**Comment by paulfantom** [comment]
I don't think it will change anything.

**Comment by ktsaou** [comment]
> I don't think it will change anything.

which one?

File: python.d/python_modules/base.py (hunk 1 above)
Comment (paulfantom) [inline python.d/python_modules/base.py]: non blocking mode per request. look at `man 2 recv`. basically `MSG_DONTWAIT`.

File: python.d/python_modules/base.py (hunk 2 above)
Comment (paulfantom) [inline python.d/python_modules/base.py]: I know.

File: python.d/python_modules/base.py (hunk 3 above)
Comment (paulfantom) [inline python.d/python_modules/base.py]: Yes, but I don't want to make any optimizations until I make it work.

**Comment by ktsaou** [comment]
Generally speaking, blocking and non-blocking should work for both services.
non-blocking (at least in C) means that the call will return immediately, without data. Then you have to poll/select/epoll for receiving data.

**Comment by ktsaou** [comment]
Since this is a threaded application, I think non-blocking is not needed. I use non-blocking only in single threaded applications, so that instead of waiting for data (send or receive), the program can serve something else.

**Comment by paulfantom** [comment]
> Generally speaking, blocking and non-blocking should work for both services.

And that's why I think it won't change anything.

**Comment by paulfantom** [comment]
I thought so too (about threaded application). But somehow it blocked everything.

**Comment by ktsaou** [comment]
Blocking mode should be very simple and straight forward. Why don't you set blocking mode and try to find what is the issue there?

**Comment by paulfantom** [comment]
The issue is `check` and `create` in `python.d.plugin`. Until `update` everything works single-threaded.

**Comment by paulfantom** [comment]
Ok, I will try to explain it a little bit more.

When starting `python.d.plugin` everything works single-threaded. Jobs are initialized, then there is one loop to check if every job can be run (`check` method). Next charts are created (`create`). Then every job is starting its own thread (in `update` method by `job.start()`). This way at the beginning I can set every variable job needs and then "fire and forget".

Problem is that `check` and `create` implemented in `SimpleService` or any other `*Service` class execute `_get_data()` method.
In case of SocketService-based modules, `_get_data()` is implemented in `Socket` class, but it executes `_get_raw_data()` from `SocketService`.

This way if I have blocking `_get_raw_data()`, it will hang whole `python.d.plugin` in `check` method.

**Comment by paulfantom** [comment]
I know this is a design flaw, but it simplifies a lot.

**Comment by paulfantom** [comment]
Also the only method that worked for every `SocketService` was slow and needed to time out.

**Comment by ktsaou** [comment]
So what you say is that you start single threaded in non-blocking mode and then you spawn threads for data collection in blocking mode.

I guess you have 2 solutions:

1. Made it completely single threaded non-blocking. If you can do this, it is the optimal. However, I think it will be complicated and as modules are will be added a lot of work will be required for each.

2. Made it multi-threaded from the beginning. Spawn threads for each job defined and let the threads that do not work exit. Then you will only need to solve the problem of mutual exclusion, which seems a lot simpler.

My 2 cents is that requiring from all modules to work in both blocking and non-blocking mode is a waste of effort (your effort).

**Comment by ktsaou** [comment]
In node.js where the thread is not a real thread, I used parallel execution of check(), but I waited for all the checks to be made. Then I started the jobs.

Something like: spawn threads in blocking mode that will just perform the check() and exit. When all complete, apply mutual exclusion and start the jobs.

This has the problem that if a single check() takes too long, all the jobs will be delayed. But the delay is not the sum of all the delays of all jobs. It is equal to the longest single job delay.

**Comment by paulfantom** [comment]
> Made it multi-threaded from the beginning. Spawn threads for each job defined and let the threads that do not work exit. Then you will only need to solve the problem of mutual exclusion, which seems a lot simpler.

This would be good, but it needs a lot of code refactoring. Every module needs to be changed.

Currently every module is a child of `BaseService` which is based on `threading.Thread`, and it should be changed if I want complete multi-threading (which is not exactly multithreading because of python GIL).

**Comment by paulfantom** [comment]
I think I will revert to old data gathering for `SocketService` and implement a different one for `redis.chart.py`. Currently this will be the simplest.

**Comment by paulfantom** [comment]
I think I got it working. There might be some issues with timeout, but I think this is the best I could do.

Could you test it when you find some time?

**Comment by ktsaou** [comment]
yes, both work.
merge?

**Comment by paulfantom** [comment]
Yes!
At last!
:tada: :tada: :tada:

**Comment by paulfantom** [comment]
Stop

**Comment by ktsaou** [comment]
ok

**Comment by ktsaou** [comment]
merge?

**Comment by paulfantom** [comment]
Yes you can.

**Comment by paulfantom** [comment]
Ok, no more commits today.

**Comment by ktsaou** [comment]
ok. Good night to you too. merged it.

**Pull request merged**